File: src/problem_005.py
# Analytical solution to a least common multiple problem:
# - find all prime factors for all values
# - get all factors with the highest exponent found

from functools import reduce
import operator

def get_prime_factors(value):
    factors = {}
    n = 2
    while value > 1:
        if value % n == 0:
            factors[n] = factors.get(n, 0) + 1
            value //= n
        else:
            n += 1
    return factors

def combine_dicts(factors1, factors2):
    for k, v in factors2.items():
        if factors1.get(k, 0) < v:
            factors1[k] = v

# accepts input n >= 2
def euler_005(n):
    factors = {}
    for v in range(2, n + 1):
        factors_v = get_prime_factors(v)
        combine_dicts(factors, factors_v)
    
    # compute result
    return reduce(operator.mul, [k ** v for k, v in factors.items()])

if __name__ == '__main__':
    print(euler_005(20))

File: src/go/problem_005.go
package main

import (
	"fmt"
	"math"
)

func getFactors(n int) map[int]int {
	m := make(map[int]int)
	i := 2
	for n > 1 {
		if n % i == 0 {
			n /= i
			m[i]++
		} else {
			i++
		}
	}
	return m
}

// return least common multiple of all values from 1 to bound
func Lcm(bound int) int {
	factors := make(map[int]int)
	for i := 2; i <= bound; i++ {
		for k,v := range getFactors(i) {
			if v > factors[k] {
				factors[k] = v
			}
		}
	}
	prod := 1.0
	for k, v := range factors {
		prod *= math.Pow(float64(k), float64(v))
	}
	return int(prod)
}

func main() {
	fmt.Println(Lcm(10))
	fmt.Println(Lcm(20))
}
